add xmpp backend

danryder opened the pull request.

dschep commented on ntfy/backends/xmpp.py:
```diff
@@ -0,0 +1,97 @@
+
+    ssl_version = ssl_version or config.get('ssl_version', None)
+    if ssl_version:
+        xmpp_bot.ssl_version = literal_eval(ssl_version)
```
Clever use of `literal_eval`. Never thought of that before. I dig it :smile:

danryder replied:
ah yes @dwyerk was where I picked up ast.literal_eval() as a safer way to eval() stuff like this, which should not um... execute

danryder replied:
btw who has since gone on to disavow such hackery as eval of any kind on config.  I suppose it is heavy handed & I can certainly split server into server-host and server-port, int()ing the port in the backend code if need be.  Your call.

dschep commented on ntfy/backends/xmpp.py:
```diff
@@ -0,0 +1,97 @@
+    # xmpp_bot.register_plugin('xep_0199') # XMPP Ping
+
+    path_to_certs = path_to_certs or config.get('path_to_certs', None)
+    if path_to_certs and os.path.isdir(path_to_certs):
```
Should we raise an exception if it's supplied and isn't a directory?

danryder replied:
yeah probably, though to be clear I didn't even test this functionality at all.  Perhaps the library would support a single cert path if you only have one.  I don't even know ;)

dschep commented on ntfy/backends/xmpp.py:
```diff
@@ -0,0 +1,97 @@
+        xmpp_bot.ssl_version = literal_eval(ssl_version)
+
+    # Connect to the XMPP server and start processing XMPP stanzas.
+    manual_server = manual_server or config.get('manual_server', None)
```
I assume it's "manual" because SleekXMPP will otherwise figure it out from the JID, yes? I'm ok with calling it just "server" since it's optional and documenting that your JID is used if you don't pass server.

danryder replied:
yep

dschep commented on ntfy/backends/xmpp.py:
```diff
@@ -0,0 +1,97 @@
+
+    # Connect to the XMPP server and start processing XMPP stanzas.
+    manual_server = manual_server or config.get('manual_server', None)
+    if manual_server:
```
I like using this pattern(and in more complicated cases, build up the args you use * on in a couple lines) in a case like this:
```
    connected = xmpp_bot.connect(*([manual_server] if manual_server else []))
```

danryder replied:
Neat I was wondering how to do that without passing None in, never thought to use *[]

dschep replied:
Ah, just delete it then :stuck_out_tongue:

dschep commented:
Good work! Could you please:
 * Add SleekXMPP to the dependencies: https://github.com/dschep/ntfy/blob/master/setup.py#L4
 * Add XMPP to the list of supported backends: https://github.com/dschep/ntfy/blob/master/README.rst#supported
 * See line comments for nit picks, mostly pep8

dschep commented:
I think #8 could help with the fact you have many options that you seem to want to allow uses to configure either via the config or the cli.

danryder commented:
I haven't ever had to create setup.py before, but it would appear that adding sleekxmpp to deps would prevent it from installing without that, which is only needed if you want to use that backend.

At any rate I've made these changes in my local branch (and commented on your pull request for unifying config options).  Not sure what needs to happen next as I've not done pull requests before, so I've held off even pushing my changes to github in case the right answer is you merge your pull request and then I rebase against the modified upstream and resolve whatever conflicts

danryder commented:
Rebased in your updates.  Also updated the new backend to just take kwargs, and killed literal_eval there (split hostname and port into 2 optional params, tho port is ignored without hostname).  Killed ssl_version option but added one for debugging (the sleekxmpp library can be pretty verbose)

danryder commented:
hmm not sure what it even means to close the pull request.  undoing that

dschep commented on ntfy/backends/xmpp.py:
```diff
@@ -0,0 +1,81 @@
+    """
+
+    def __init__(self, jid, password, recipient, title, message, mtype=None):
+        sleekxmpp.ClientXMPP.__init__(self, jid, password)
```
This should probably be using `super()`:
```
super(NtfySendMsgBot,self).__init__(jid, password)
```

danryder replied:
done, force pushed - latest is a0e02bd2e
